WMF Framework/CollectionViewCellActionsView.swift

```diff
@@ -0,0 +1,193 @@
+        }
+    }
+}
+
+public protocol SwipeableDelegate: NSObjectProtocol {
+    func didOpenActionPane(_ didOpen: Bool, at: IndexPath)
+    func didTapSave(at indexPath: IndexPath)
+    func didTapUnsave(at indexPath: IndexPath)
+    func didTapShare(at indexPath: IndexPath)
+    func didTapDelete(at indexPath: IndexPath)
+    func isArticleSaved(at indexPath: IndexPath) -> Bool
+}
+
+public class CollectionViewCellActionsView: UIView {
+    
+    var cell: ArticleCollectionViewCell
+    var maximumWidth: CGFloat = 0
+    public var theme = Theme.standard
+    
+    var actions: [CollectionViewCellAction] = [] {
+        didSet {
+            createSubviews(for: self.actions)
+        }
+    }
+    
+    var swipeType: CollectionViewCellSwipeType = .none {
+        didSet {
+            swipeType == .primary ? layoutPrimaryActions() : layoutSecondaryActions()
+        }
+    }
+    
+    init(frame: CGRect, cell: ArticleCollectionViewCell) {
+        self.cell = cell
+        super.init(frame: frame)
+        
+        self.isUserInteractionEnabled = false
+    }
+    
+    func createSubviews(for actions: [CollectionViewCellAction]) {
+        
+        for view in self.subviews {
+            view.removeFromSuperview()
+        }
+        
+        var maxButtonWidth: CGFloat = 0
+        
+        for (index, action) in actions.enumerated() {
+            let button = UIButton(type: .custom)
+            button.setTitle(action.title, for: .normal)
+            button.titleLabel?.numberOfLines = 1
+            button.contentEdgeInsets = UIEdgeInsetsMake(0, 14, 0, 14)
+            button.tag = index
+            
+            var didTapButton: Selector
+            
+            switch (action.type) {
+            case .delete:
+                didTapButton = #selector(didTapDelete)
+                button.backgroundColor = theme.colors.destructive
+            case .share:
+                button.backgroundColor = theme.colors.secondaryAction
+                didTapButton = #selector(didTapShare)
+            case .save:
+                button.backgroundColor = theme.colors.link
+                didTapButton = #selector(didTapSave)
+            case .unsave:
+                button.backgroundColor = self.theme.colors.link
+                didTapButton = #selector(didTapUnsave)
+            }
+            
+            button.addTarget(self, action: didTapButton, for: .touchUpInside)
+            
+            // Wrapper around each button.
+            let wrapper = UIView(frame: .zero)
+            wrapper.clipsToBounds = true
+            wrapper.addSubview(button)
+            self.addSubview(wrapper)
+            maxButtonWidth = max(maxButtonWidth, button.intrinsicContentSize.width)
+        }
+        
+        maximumWidth = maxButtonWidth * CGFloat(self.subviews.count)
+    }
+    
+    public weak var delegate: SwipeableDelegate?
+    
+    @objc func didTapDelete() {
```

Comment:
This might benefit from being consolidated into a single method that passes the action or action type to the delegate. After consolidating, it would be easier to add new actions without adding new methods to this class and every delegate.